functions/auth.js

```diff
@@ -19,3 +30,164 @@ exports.passwordsMatch = (submittedPassword, actualPassword) =>
+
+  functions.logger.log("Received auth code:", authCode);
+  const results = await authClient.authorizationCode.getToken({
+    code: authCode,
+    redirect_uri: "http://localhost:3000/in/foo", // TODO: what do we want to use here?
+  });
+  functions.logger.log("Auth code exchange result received:", results);
+
+  const { access_token: accessToken } = results;
+
+  // Retrieve the user's I4A User ID
+  const { id: i4aUserId } = await getJson(i4aOAuthUserInfoUrl, {
+    Authorization: `Bearer ${accessToken}`,
+  });
+
+  functions.logger.log("I4A User ID:", i4aUserId);
+
+  if (!i4aUserId) {
+    // TODO: redirect to some kind of '500 error' page?
+    res.redirect("/in/TODO/error");
+    return;
+  }
+
+  // TODO: configure this in cloud config and/or firestore or similar
+  const meetingIdsToCheck = [131, 136];
+  const eventIdsToCheck = [504, 506, 554];
+
+  const checkedMeetingResult = await postJson(i4aGetUserMeetingInfoUrl, {
+    apiKey: i4aApiKey,
+    ams_id: i4aUserId,
+  });
+
+  functions.logger.log("Checked Meeting Result:", checkedMeetingResult);
+
+  const {
+    email: emailRaw,
+    meeting_ids: meetingIds = [],
+    event_ids: eventIds = [],
+  } = checkedMeetingResult;
+
+  const registeredMeetings = meetingIds.filter((meetingId) =>
+    meetingIdsToCheck.includes(meetingId)
+  );
+
+  const registeredEvents = eventIds.filter((eventId) =>
+    eventIdsToCheck.includes(eventId)
+  );
+
+  const isRegistered =
+    registeredMeetings.length > 0 && registeredEvents.length > 0;
+
+  functions.logger.log("Registered:", {
+    isRegistered,
+    registeredMeetings,
+    registeredEvents,
+  });
+
+  if (!isRegistered || !emailRaw) {
+    // TODO: redirect to some kind of 'not allowed' page
+    res.redirect("/in/TODO/not-registered");
+    return;
+  }
+
+  const email = emailRaw.toLowerCase().trim();
+
+  // Lookup the existing user by their email, or create them if they don't already exist
+  const userRecord = await admin
+    .auth()
+    .getUserByEmail(email)
+    .catch((error) => {
+      // If user doesn't exist then create them
+      if (error.code === "auth/user-not-found") {
+        functions.logger.log(
+          "Existing user not found, creating new user:",
+          email
+        );
+
+        // We explicitly don't set a password here, which should prevent signing in that way (until the user resets their password to create one)
+        // We also explicitly aren't creating the user's profile here, which will let them configure it in the normal way when they first sign in
+        return admin.auth().createUser({ email });
+      }
+      throw error;
+    });
+
+  functions.logger.log("User:", {
+    i4aUserId,
+    userId: userRecord.uid,
+    email: userRecord.email,
+  });
+
+  // Create a custom token for the frontend to use to sign into firebase auth as this user
+  const customToken = await admin.auth().createCustomToken(userRecord.uid, {
+    registeredMeetings,
+  });
```

Comment:
Do we want to include `registeredMeetings` in the custom claims here, is it relevant? If we do keep it, we should also add `registeredEvents`

Reply:
Fixed this, not sure of the commit off the top of my head. Basically moved the claims to the user itself rather than the custom token, and added both of them.